### DISCUSSION TITLE
how to read env dynamically after next start

### DISCUSSION THREAD

**Discussion by seognil** [vercel/next.js#14966] [Help]
how to read `env` dynamically after build and start, and then pass them to the client-side

--- 

Here's the case, I'm using docker. in Dockerfile, it's like:

```
RUN npm run build

CMD npm start
```

I want to deploy the single image to various servers (test, www, etc), with different env (API_URL, CDN_URL, etc)
So, the progress is: next build (docker build image) -> changing env (docker run with args) -> next start (CMD in docker run)

And it's not working as I thought (even without docker).

I tried different ways by what I knew: tweaking next.config.js, change env name to NEXT_PUBLIC_, blah blah. But still can't let it happen.

---

Finally found a tricky way,
read `process.env.XXX` in `getServerSideProps()`  then pass them to Comp as props. That works well.

Is there any better solution?

**Comment by jamesmosier** [comment]
You could look at `serverRuntimeConfig` for dynamically setting variables: https://nextjs.org/docs/api-reference/next.config.js/runtime-configuration

These would live in your `next.config.js` and you could swap them out anytime.

**Comment by seognil** [comment]
Thanks, I've tried this demo: https://github.com/vercel/next.js/blob/master/examples/with-universal-configuration-runtime/next.config.js

---

- dev
- change env
- dev

worked

---

- build
- change env
- start

not worked

---

I'm tracing the next script code now,
saw `next.config.js` is reloaded every time run start, but the value in pages won't changing
didn't figure out what's going on yet...

**Comment by millievn** [comment]
just use [getConfig](https://nextjs.org/docs/api-reference/next.config.js/runtime-configuration) and i create a [demo](https://github.com/qiwang97/nextjs-discussions-14966)

**Comment by seognil** [comment]
Still not working as I expected

## Simple explain

Why this?

Think about one single docker image with an nextjs app burned in,
and publish to multiple servers with k8s and different env configs blahblahblah.

## What's wrong?

based on  @qiwang97 's demo

---

1st run (dev)

```sh
export BASE_URL_ONLY='dev_only'; export BASE_URL_BOTH='dev_both';
yarn dev;
```

rendered (**OK**)

```
BASE_URL_ONLY:
BASE_URL_BOTH: dev_both
```

---

2nd run (dev)

```sh
export BASE_URL_ONLY='dev_only_again'; export BASE_URL_BOTH='dev_both_again';
yarn dev;
```

rendered (**OK**)

```
BASE_URL_ONLY:
BASE_URL_BOTH: dev_both_again
```

---

3rd run (prod)

```sh
export BASE_URL_ONLY='build_only'; export BASE_URL_BOTH='build_both';
yarn build;
yarn start;
```

rendered (**OK**)

```
BASE_URL_ONLY:
BASE_URL_BOTH: build_both
```

---

4th run (prod)
(just change the env and start, don't build again, (as we can't rebuild the app after the docker image created))

```sh
export BASE_URL_ONLY='build_only_again'; export BASE_URL_BOTH='build_both_again';
yarn start;
```

rendered (**NOT OK**)
variables not changing

```
BASE_URL_ONLY:
BASE_URL_BOTH: build_both
```

(so we can't achieve that with one single docker image (in this way))

**Reply by millievn** [reply]
i just know what you really want now and i did  [this](https://github.com/vercel/next.js/issues/7844#issuecomment-650728953)  as temporary solution. Maybe you can try it.

**Reply by nilevia** [reply]
I have same goals with you, one single docker image with two environtment (staging and production). I tried to achive this by move yarn build into entrypoint. But this cause another problem. My website got 502 for few minutes after deployed (need time to execute build) then it working fine.

**Comment by praveen-tripathi** [comment]
I'm also looking for a solution for exactly the same case. One build, multiple environments(server). Any thoughts?

**Comment by Mistra** [comment]
This is a pain.
As a workaround I'm picking the variable on the server component and I shove it down to the client components that need it. It's ugly... but it works.

I'm thinking to implement on my own a more structured way to avoid the hack but it should be the framework to provide the functionality.

**Comment by Aiosa** [comment]
There is another way, you can use init container that builds the code before the main container, share the build mount with the production container, and swap after build from a custom ENV state. I do this in kubernetes. Honestly, this is truly retarded behavior. I came here only because now I am forced to pull from private repository so I wanted to create a static image to avoid dealing with permissions, and remembered this. Yuck.
